### ISSUE TITLE
Improve display logic of Express Payment buttons in Checkout block

### ISSUE THREAD

**Issue by nielslange** [woothemes/woocommerce#42464]
## Describe the bug

In woocommerce/woocommerce-blocks#8601, we modified the presentation of express payment buttons to appear in a three-column grid. In https://github.com/woocommerce/woocommerce/issues/42464, @pierorocca mentioned that no more than three express payment buttons should be displayed.

This issue aims to enhance the display logic so that a maximum of three express payment buttons are visible. If a merchant has set up more than three express payment methods, the customer should see three express payment buttons along with an option to view the remaining express payment methods.

**Comment by stephendharmadi** [comment]
Thanks for pinging me on this!

I am working on these payment button guidelines, and soon to be finalized!

For both horizontal and vertical layouts, we will place 3 express payment buttons maximum in the same row. If merchants provide more than 3 express payments, then we will display the CTA "More express payments" which displays the modal view when it's clicked.

### Horizontal layout - 3 buttons
![Horizontal layout - 3 Buttons](https://user-images.githubusercontent.com/110893979/221863289-80672ef7-d6f6-47f7-aa57-b34dc147210c.png)

### Horizontal layout - more than 3 buttons
![Horizontal layout - more than 3 Buttons](https://user-images.githubusercontent.com/110893979/221863606-1687d205-a052-41f7-aa72-dc43311e9d9f.png)

### Horizontal layout - Display 6 buttons
![Horizontal layout - Display 6 Buttons](https://user-images.githubusercontent.com/110893979/221863876-7c210b09-7b21-4131-8e63-8754da413715.png)

### Horizontal layout - Display 9 buttons
![Horizontal layout - Display 9 Buttons](https://user-images.githubusercontent.com/110893979/221866117-35e95181-37f0-4948-a1a0-5adbe0bb57c4.png)

You can check out the guidelines, including the vertical layout in this [Figma file](https://www.figma.com/file/qNVPHJXgRkftiSR7j2XOlj/Express-Payments-Button-Guidelines?node-id=3%3A8599&t=eaBZ9Esj3MaTuNBd-4)

Let me know if you have any questions.

**Comment by nerrad** [comment]
Some questions related to the design for the modal:

- Who has control over which express payment methods overflow into the modal? How is this controlled?
- Instead of a modal, have you explored the button triggering an expansion of the existing container? Put another way, why is a modal the option here?
- Should the modal display all express payment method options instead of just the overflowing options? That would save an additional click if the shopper wanted to use one of the original three.

**Comment by stephendharmadi** [comment]
Hey @nerrad, to answer your questions.

> Who has control over which express payment methods overflow into the modal? How is this controlled?

Definitely merchants! They can set the priority order of payment methods to be displayed on checkout. Also, I think we currently have the functionality to do that in "All payment methods."

<img width="2312" alt="Screenshot 2023-03-01 at 10 53 52" src="https://user-images.githubusercontent.com/110893979/222041258-98216a6d-b8a6-44bd-b714-3c5e2054baad.png">

> Instead of a modal, have you explored the button triggering an expansion of the existing container? Put another way, why is a modal the option here?

Yes, I have considered this solution. Although it can simplify the first view of checkout, when it's expanded, the problem of overcrowded buttons still exists, which could take up the entire checkout page (specifically for vertical layout like example below).

![Vertical layout - more than 3 Buttons](https://user-images.githubusercontent.com/110893979/222042931-3a0f0f69-7490-4770-85ca-7f87266b1829.png)

![Mitigation 2 - Horizontal layout - more than 3 Buttons](https://user-images.githubusercontent.com/110893979/222043279-2f416a8d-3ee0-419d-b181-f650a38a26c0.png)

Probably what I could think of all possible solutions:

- Alternative 1
Limiting the number of express payments that merchants can enable at the same time (let's say 5). The downside is we would limit merchants' opportunity to use any payment available in the market. Something that we can validate from merchants is how many payments they wish to have on their checkout. cc @pierorocca if you have any insights about this.

- Alternative 2
Do not limit express payments that merchants can enable. Automatically using the horizontal layout if merchants use more than 5 express payments.

- Alternative 3
The modal display approach can provide flexibility to not limit express payments merchants can enable, and at the same time provides a simple and clean checkout view.

> Should the modal display all express payment method options instead of just the overflowing options? That would save an additional click if the shopper wanted to use one of the original three.

It's a good point. But I couldn't find why merchants need to click the "more express payments" CTA if they intended to select the original three which are visible in the first view.

**Comment by elizaan36** [comment]
> They can set the priority order of payment methods to be displayed on checkout. Also, I think we currently have the functionality to do that in "All payment methods."

@nerrad For Woo admin, this update is also part of a design proposal to update Payments settings to allow for easier reordering of express payment methods. More details in `p6riRB-8kS`

However, it could start as a blocks-specific update where the reordering is only available in blocks settings and only applies to Blocks. Stephen has an example of this in his explorations. 
<img width="280" alt="image" src="https://user-images.githubusercontent.com/5121465/222526476-7543cd11-c557-4feb-9e4e-b0d0e9a34599.png">
 
I would definitely go with Alternative 2 or 3 over limiting the number of express methods at checkout. I do like how the modal keeps the checkout page nice and clean.

**Comment by nielslange** [comment]
Closing this issue due to woocommerce/woocommerce-blocks#8601.

**Comment by pierorocca** [comment]
9 buttons? I distinctly recall Matt M calling out 6 as being ridiculous. I'd recommend 3 max and then either allow them to cycle horizontally on a carousel and/or build out some sort of predictive capability similar to what Stripe has done with UPE.

In any case based on studies of button performance, merchants should be encouraged to pick 1 or 2 EPMs and the rest offered as standard payment methods.

**Comment by nielslange** [comment]
@pierorocca I closed the issue as I was thinking the primary intention of this issue was to add the spacing between the express payment buttons, which was fixed in woocommerce/woocommerce-blocks#8601. By closing this issue, it was not my intention to show an unlimited number of express payment buttons in a 3-column grid. That said, I'll reopen the issue and update the description so that it's clear that this issue should ensure that no more than 3 express payment buttons should be visible.

**Comment by ralucaStan** [comment]
@stephendharmadi could you point us to the final design for this behavior? Thank you so much

> This issue aims to enhance the display logic so that a maximum of three express payment buttons are visible. If a merchant has set up more than three express payment methods, the customer should see three express payment buttons along with an option to view the remaining express payment methods.

**Comment by stephendharmadi** [comment]
Hi @ralucaStan, I'm revisiting the WooPay button guidelines to establish the guidelines and hopefully publish it as an open source. For the express payments, I have just explored it this week and establish the new button sizes and layout with other payments. To summarize, the changes cover:
- Enlarge the default size (previously 38px) to 48px. You could refer to [the specs here](https://www.figma.com/file/qNVPHJXgRkftiSR7j2XOlj/Payment-Element-Guidelines?type=design&node-id=174-15989&t=ddA2r94xyH7Ap1WU-4)
- Establish [the button structure for different sizes](https://www.figma.com/file/qNVPHJXgRkftiSR7j2XOlj/Payment-Element-Guidelines?type=design&node-id=174-11491&t=ddA2r94xyH7Ap1WU-4), layouts, and button types (with label or icon only). You can play with [the Figma mockup](https://www.figma.com/file/qNVPHJXgRkftiSR7j2XOlj/Payment-Element-Guidelines?type=design&node-id=227-11289&t=ddA2r94xyH7Ap1WU-4) and change the variables to easily explore (example below)


https://github.com/woocommerce/woocommerce-blocks/assets/110893979/dced8fe2-a8dd-42a1-94f0-e37a84aa48f4

**Comment by elizaan36** [comment]
Thanks for sharing this exploration @stephendharmadi. Looks good however let's collaborate with Shaun on this. There's starting to be some clarity on the solution based on the work posted here: `pdnLyh-3L2`

**Comment by ralucaStan** [comment]
This can be closed following the work in https://github.com/woocommerce/woocommerce/pull/50791; 
https://developer.woocommerce.com/2024/10/28/more-options-coming-to-the-express-checkout/

**Comment by ralucaStan** [comment]
@alexflorisca the feature has been released as beta. We should start considering removing this from the beta. Could you start a developer discussion on the topic?

**Comment by pierorocca** [comment]
> @alexflorisca the feature has been released as beta. We should start considering removing this from the beta. Could you start a developer discussion on the topic?

We'll need to align on this internally first @ralucaStan as the settings are currently fragmented between back and front-ends.  It's a similar situation to Add to Cart settings where we need to better coordinate settings that can co-exist in wp-admin and FSE.

**Comment by alexflorisca** [comment]
Yeah I know you mentioned this before Pierro, let's have a think about that. I think the challenge is that the settings in the payment gateway settings are owned by the plugin, and the settings in the block are owned by us. If you have any more thoughts on how we could do this, I'm all ears. I think one of the ideas was to at least sync SOME of the settings like button size, border radius, etc with the plugins that support this and vice versa.

**Comment by pierorocca** [comment]
I think it needs to be a data store approach since this needs state management. An override is a powerful tool. It may be better to move this out of the block and into the new settings pages, which aims to consolidate settings for all registered gateways, all registered payment methods. I could see it working much better there than in the block. I'll ping you the designs.

**Comment by alexflorisca** [comment]
Sounds good, I guess we should revisit this when the designs are implemented?

**Comment by pierorocca** [comment]
I think so since it achieves the same thing in a more holistic way.
